pycbc/workflow/dq.py

```diff
@@ -22,183 +22,122 @@
+        node.add_input_opt('--template-bins-file', template_bins_file)
         node.add_input_opt('--trig-file', trig_file)
-        node.add_input_list_opt('--dq-file', dq_files)
-        node.new_output_file_opt(workflow.analysis_time,'.hdf',
+        node.add_input_opt('--flag-file', flag_file)
+        node.add_opt('--flag-name', flag_name)
+        node.add_input_opt('--analysis-segment-file', analysis_segment_file)
+        node.add_opt('--analysis-segment-name', analysis_segment_name)
+        node.new_output_file_opt(workflow.analysis_time, '.hdf',
                                  '--output-file')
         return node
 
-class PyCBCCalculateDQFlagExecutable(Executable):
-    # current_retention_level = Executable.ALL_TRIGGERS
-    current_retention_level = Executable.MERGED_TRIGGERS
-
-    def create_node(self, workflow, segment, dq_file, flag):
-        node = Node(self)
-        # Executable objects are initialized with ifo information
-        start = int(segment[0])
-        end = int(segment[1])
-        node.add_opt('--ifo', self.ifo_string)
-        node.add_opt('--flag', flag)
-        node.add_opt('--gps-start-time', start)
-        node.add_opt('--gps-end-time', end)
-        node.add_input_opt('--dq-segments', dq_file)
-        node.new_output_file_opt(segment, '.hdf',
-                                 '--output-file')
-        return node
 
-def setup_dq_reranking(workflow, dq_label, insps, bank,
-                        segs, analyzable_file, dq_file,
-                        output_dir=None, tags=None):
+def setup_dq_reranking(workflow, insps, bank,
+                       analyzable_seg_file,
+                       analyzable_name,
+                       dq_seg_file,
+                       output_dir=None, tags=None):
+    logging.info("Setting up dq reranking")
     make_analysis_dir(output_dir)
-    output = FileList()
-    if tags:
+    output_files = FileList()
+    output_labels = []
+    if tags is None:
+        tags = []
+
+    dq_labels = workflow.cp.get_subsections('workflow-data_quality')
+    dq_labels = numpy.array(dq_labels)
+
+    dq_types = numpy.array([workflow.cp.get_opt_tags(
+            'workflow-data_quality', 'dq-type', [dq_label])
+            for dq_label in dq_labels])
+    dq_ifos = numpy.array([workflow.cp.get_opt_tags(
+            'workflow-data_quality', 'dq-ifo', [dq_label])
+            for dq_label in dq_labels])
+    dq_names = numpy.array([workflow.cp.get_opt_tags(
+            'workflow-data_quality', 'dq-name', [dq_label])
+            for dq_label in dq_labels])
+
+    ifos = set(dq_ifos)
+
+    for ifo in ifos:
+        # FIXME : make this able to take multiple dq files per ifo
+
+        # get the dq label, type, and name for this ifo
+        mask = (dq_ifos == ifo)
+        assert numpy.sum(mask) > 0, f"Received no dq files for {ifo}"
+        assert numpy.sum(mask) < 2, f"Received more than one dq file for {ifo}"
+        dq_label = str(dq_labels[mask][0])
```

Comment:
comment on why `str` is needed ?

Reply:
Numpy converts it to some kind of numpy specific string format that pegasus doesn't like. If I don't use arrays, this won't be necessary